### ISSUE TITLE
DDB Stream Lambda creation fails due to IAM policy not being updated first

### ISSUE THREAD

**Issue by flyinbutrs** [serverless/serverless#3997]
# This is a Bug Report

## Description

* What went wrong?

While updating an existing deployment with a new DDB Stream lambda, while simultaneously updating the IAM policy for the role and the DDB table to add the stream, the AWS::Lambda::EventSourceMapping fails to create with the following error:

````
Cannot access stream arn:aws:dynamodb:us-west-2:123456789012:table/logger-api-gateway-stage-2/stream/2017-07-26T18:04:04.509. Please ensure the role can perform the GetRecords, GetShardIterator, DescribeStream, and ListStreams Actions on your stream in IAM.
````

These permissions are due to be added in the update, but because the AWS::Lambda::EventSourceMapping is not set to depend on the IAM role, the updates don't happen in the correct sequence.

* What did you expect should have happened?

AWS::Lambda::EventSourceMapping's that are for DynamoDB Streams should have "DependsOn": "IamRoleLambdaExecution"

* What was the config you used?

````
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:DescribeStream
        - dynamodb:GetItem
        - dynamodb:GetRecords
        - dynamodb:GetShardIterator
        - dynamodb:ListStreams
        - dynamodb:PutItem
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:UpdateItem
      Resource:
        - Fn::GetAtt:
          - LogsTable
          - StreamArn
functions:
  bus:
    description: Processes changes from DynamoDB and relays as appropriate
    handler: logs/bus.process
    events:
      - stream:
          type: dynamodb
          arn:
            Fn::GetAtt:
              - LogsTable
              - StreamArn
````

* What stacktrace or error message from your provider did you see?

````
  Serverless Error ---------------------------------------
 
  An error occurred while provisioning your stack: BusEventSourceMappingDynamodbLogsTable - Cannot access stream arn:aws:dynamodb:us-west-2:012345678901:table/logger-api-gateway-stage-2/stream/2017-07-26T18:04:04.509. Please ensure the role can perform the GetRecords, GetShardIterator, DescribeStream, and ListStreams Actions on your stream in IAM..
 ````

## Additional Data

* ***Serverless Framework Version you're using***:

1.16.1

* ***Operating System***:

ubuntu 16.04

**Comment by flyinbutrs** [comment]
I suspect that in this line, it needs to be IamRoleLambdaExecutionPolicy, rather than IamRoleLambdaExecution. That way if the policy is to be updated, it'll get updated before the EventSourceMapping is created.

https://github.com/serverless/serverless/blob/aebce86ab20faaf95db32e037990de81f28e8dda/lib/plugins/aws/package/compile/events/stream/index.js#L122

**Comment by alexcasalboni** [comment]
Hi @flyinbutrs, I am experiencing the same problem with Kinesis Streams.

I get a similar error: 

> An error occurred: MyFunctionEventSourceMappingMyKinesisStream - Cannot access stream arn:aws:kinesis:us-east-1:XXX:stream/YYY. Please ensure the role can perform the GetRecords, GetShardIterator, DescribeStream, and ListStreams Actions on your stream in IAM.

Have you tested your suggested fix related to `dependsOn`?

CC @pmuens you can reproduce the bug by starting with a yaml configuration with no `iamRoleStatements` and no `events`, such as:

```yaml
service: xxx
provider:
  name: aws
  runtime: python2.7
functions:
  myFunction:
    handler: handler.my_handler
```

You deploy it, and then you update the config to:

```yaml
service: xxx
provider:
  name: aws
  runtime: python2.7
  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - kinesis:GetRecords
        - kinesis:GetShardIterator
        - kinesis:DescribeStream
        - kinesis:ListStreams
      Resource:
        - Fn::GetAtt:
          - MyKinesisStream
          - Arn

functions:
  myFunction:
    handler: handler.my_handler
    events:
      - stream:
          type: kinesis
          batchSize: 100
          arn:
            Fn::GetAtt:
              - MyKinesisStream
              - Arn

resources:
  Resources:
    MyKinesisStream:
      Type: AWS::Kinesis::Stream
      Properties:
        ShardCount: 1
```

I believe this did not happen until a few months ago because AWS was not validating Lambda's permissions when you created a new EventSourceMapping.

I wouldn't consider the bug super critical, as everything works fine if you `sls remove` and then `sls deploy` again, but it can be quite frustrating for beginners.

**Comment by madaray** [comment]
I think this may be due to the iam definition. 

```- Effect: "Allow"
      Action:
        - "kinesis:GetRecords"
        - "kinesis:GetShardIterator"
        - "kinesis:DescribeStream"
        - "kinesis:ListStreams"
      Resource:
        - "arn:aws:kinesis...```

The above deploys but on checking this role there are issues with `kinesis:GetRecords` and `kinesis:ListStreams` being defined with a named resource. This prevents the trigger being added directly through the console also. 

Updating to this fixed it:

``` - Effect: "Allow"
      Action:
        - "kinesis:GetShardIterator"
        - "kinesis:DescribeStream"
      Resource:
        - "arn:aws:kinesis..."

    - Effect: "Allow"
      Action:
        - "kinesis:GetRecords"
        - "kinesis:ListStreams"
      Resource:
        - "*"```

**Comment by alexcasalboni** [comment]
I don't think that's the problem, because everything works fine if you
deploy the IAM role first, and then the trigger. I still think it's a
CloudFormation dependency problem.

**Comment by thundo** [comment]
Had the same problem with lambda and kinesis streams. @alexcasalboni "two-phase deploy" (first triggerless then with triggers) worked. Any fix to this issue?

**Comment by alexcasalboni** [comment]
@thundo didn't have time to investigate yet. If it's CloudFormation related as I assumed, it will take a bit of digging and testing efforts. I may have time to work on a fix next month :)

**Comment by flyinbutrs** [comment]
I'm 90% sure I pointed out exactly the fix that needs to happen, but I don't have a good example to test with at the moment, and I'm not a node dev, so I'm not inclined to start figuring out how to setup a test env to try it. But if someone is those things, I think all you need to do is change this line:

https://github.com/serverless/serverless/blob/aebce86ab20faaf95db32e037990de81f28e8dda/lib/plugins/aws/package/compile/events/stream/index.js#L122

to this:

```javascript
 let dependsOn = '"IamRoleLambdaExecutionPolicy"'; 
```

**Comment by alexcasalboni** [comment]
@flyinbutrs I have checked the code more in depth, and it looks like we can't just use "IamRoleLambdaExecutionPolicy", simply because that resource does not exist in the CloudFormation template.

In order for the `AWS::Lambda::EventSourceMapping` to depend on a `AWS::IAM::Policy`, we should change how we define the default IAM Role [here](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/lib/iam-role-lambda-execution-template.json) and separate it from the policy (which will have its own logical name).

Since the codebase assumes that the default IAM Role and its Policy are defined together (without references), we'd need to apply more than one change here and there.

**Comment by alexcasalboni** [comment]
@flyinbutrs Update: I took an hour to work on it and opened [this PR](https://github.com/serverless/serverless/pull/4427).

It works :)

Unfortunately, I haven't updated all the tests yet, so it won't be merged asap. Hopefully someone can help with that (CC @pmuens @horike37), otherwise I'll keep working on it next week.

In the meantime, you could test if it works for you as well.

**Comment by pmuens** [comment]
Seems like this was fixed on AWS end (see convo in https://github.com/serverless/serverless/pull/4427). Feel free to re-open if this is still an issue.

**Comment by patelman** [comment]
I am still getting error in CloudFormation : 

15:42:28 UTC-0400 | CREATE_FAILED | AWS::Lambda::EventSourceMapping | PutSessionFunctionStream | Cannot access stream arn:aws:dynamodb:us-west-2:xxxxxxxxxxxxx:table/xyz123-354650-ssdfe-ZNL66UUDFRHU/stream/2019-06-21T19:40:54.643. Please ensure the role can perform the GetRecords, GetShardIterator, DescribeStream, and ListStreams Actions on your stream in IAM. (Service: AWSLambda; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: ...)

**Comment by patelman** [comment]
> Seems like this was fixed on AWS end (see convo in #4427). Feel free to re-open if this is still an issue.

It is still an issue with dynamodb and Iam policy dependencies.

**Comment by ahmedrshdy** [comment]
Same issue here

**Comment by ajfisher** [comment]
Yep seeing this as well though I've used terraform to create the IAM policies and roles and this has been assigned to the policy for the role. 

This error is way less helpful for debugging the cause than a lot of the others (eg IAM permissions failures for specific capabilities) so I wonder if it's actually this that's the problem?

**Comment by joekiller** [comment]
If anyone has this problem with the  role as defined the problem here is that the "Resource" definition is incorrect to allow stream processing. In this case the example lists:

```
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:DescribeStream
        - dynamodb:GetItem
        - dynamodb:GetRecords
        - dynamodb:GetShardIterator
        - dynamodb:ListStreams
        - dynamodb:PutItem
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:UpdateItem
      Resource:
        - Fn::GetAtt:
          - LogsTable
          - StreamArn
```

when it should be:


```
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:DescribeStream
        - dynamodb:GetItem
        - dynamodb:GetRecords
        - dynamodb:GetShardIterator
        - dynamodb:ListStreams
        - dynamodb:PutItem
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:UpdateItem
      Resource:
        - Fn::GetAtt:
          - LogsTable
          - StreamArn
    - Effect: Allow
      Action:
        - dynamodb:DescribeStream
        - dynamodb:GetRecords
        - dynamodb:GetShardIterator
        - dynamodb:ListStreams
      Resource:
        - Fn::Join::
          - ""
          - - Fn::GetAtt:
                - LogsTable
                - StreamArn
            -"/stream/*"
```

See https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/using-identity-based-policies.html#access-policy-examples-for-sdk-cli.example7 for more details.

**Comment by jheising** [comment]
Not sure if this helps, but I think this is actually a problem with the error message— it doesn't list all the permissions you actually need. If you look at https://docs.amazonaws.cn/en_us/lambda/latest/dg/with-kinesis.html under "Execution role permissions" it says you need more permissions. I've updated my serverless.yml with the following and it seems to work:

```
      - Effect: "Allow"
        Action:
          - kinesis:DescribeStream
          - kinesis:DescribeStreamSummary
          - kinesis:GetRecords
          - kinesis:GetShardIterator
          - kinesis:ListShards
          - kinesis:ListStreams
          - kinesis:SubscribeToShard
        Resource: some_resource_here
```

**Comment by pgrzesik** [comment]
Hello @jheising - could you share a reproducible example where the default policies are not enough? We should be adding all of those actions, but maybe there's a case that we're processing incorrectly. Also, what version of the Framework are you currently using?

**Comment by bc4253** [comment]
I just ran into a problem adding a Kinesis data stream trigger to a newly created lambda function.  Adding the trigger without having first added anything to the default IAM policy gave an error message saying GetRecords, GetShardIterator, DescribeStream, ListShards, and ListStreams Actions permissions needed to be added.   Adding only these permissions did not let me create the Lambda Trigger.  After adding SubscribeToShard, GetShardIterator and DescribeStreamSummary permissions, I was able to add the trigger.
